_posts/2024-12-02-safety-of-methods-for-numeric-primitive-types.md

```diff
@@ -0,0 +1,395 @@
+
+```rust
+#[cfg(kani)]
+mod verify {
+    use super::*;
+
+    // Generate harnesses for methods involving two operands
+    // `type`: Integer type
+    // `method`: Method to verify
+    // `harness_name`: A unique identifier for a harness
+    macro_rules! generate_unchecked_math_harness {
+        ($type:ty, $method:ident, $harness_name:ident) => {
+            #[kani::proof_for_contract($type::$method)]
+            pub fn $harness_name() {
+                let num1: $type = kani::any::<$type>(); // (1)
+                let num2: $type = kani::any::<$type>(); // (1)
+                // (2)
+                unsafe { num1.$method(num2); } // (3)
+                // (4)
+            }
+        }
+    }
+
+    generate_unchecked_math_harness!(i8, unchecked_add, unchecked_add_i8);
+    generate_unchecked_math_harness!(i16, unchecked_add, unchecked_add_i16);
+    // ... Repeat for other integer types (32/64/128-bit/architecture-dependent signed/unsigned integers)
+}
+```
+
+A harness contains several parts:
+
+1. [**Symbolic values**](https://model-checking.github.io/kani/tutorial-nondeterministic-variables.html): `num1` and `num2` are symbolic values generated using `kani::any()`. Kani employs symbolic execution to explore a wide range of input possibilities systematically.
+2. **Assumptions**: Kani expands `#[requires(precondition)]` to `kani::assume(precondition)` and inserts this assumption in the harness before the unsafe function call. This assumption ensures that all generated values respect the preconditions of `unchecked_add`. If there are additional assumptions not captured by function contracts, you can specify them manually as well. You can find further details in [Kani's blog about function contracts](https://model-checking.github.io/kani-verifier-blog/2024/01/29/function-contracts.html).
+3. **Unsafe Execution**: The invocation of `unchecked_add` within an unsafe block for verification. [`unsafe`](https://doc.rust-lang.org/book/ch19-01-unsafe-rust.html) code blocks enable unsafe Rust features, such as calling unsafe functions or dereferencing raw pointers.
+4. **Assertions**: Similar to assumptions, Kani expands `#[ensures(postcondition)]`to `kani::assert(postcondition)` and inserts in the harness after the unsafe function call. This assertion checks if the function behaves as expected (e.g. returning an expected result) or if certain safety invariants hold after function call.
+
+**Summary**: Kani verifies a function's contract by assuming the specified preconditions, executing the function within these constraints, and asserting that the postconditions are satisfied. During this process, Kani ensures that the unsafe code adheres to Rust's safety guarantees and does not trigger undefined behavior.
```

Comment:
@carolynzech requesting review on the summary part.

Reply:
Looks great!